src/main/scala/beam/agentsim/agents/choice/logit/MultinomialLogit.scala

```diff
@@ -36,35 +38,39 @@ class MultinomialLogit[A, T](
     if (alternatives.isEmpty) None
     else {
 
-      // one-pass evaluation and sorting in descending order by alternative utilities
-      val altsWithUtilitySortedDesc: Iterable[(A, Double)] =
-        alternatives.foldLeft(SortedSet.empty[(A, Double)](Ordering.by { -_._2 })) {
-          case (set, (alt, attributes)) =>
+      // evaluate utility of alternatives
+      val altsWithUtility: Iterable[AlternativeWithUtility[A]] =
+        alternatives.foldLeft(List.empty[AlternativeWithUtility[A]]) {
+          case (accumulator, (alt, attributes)) =>
             getUtilityOfAlternative(alt, attributes) match {
-              case None => set
+              case None => accumulator
               case Some(thisUtility) =>
-                set + ((alt, math.exp(thisUtility)))
+                if (thisUtility == Double.PositiveInfinity) {
+                  // place on tail of list, allowing us to short-circuit the sampling in next step
+                  accumulator :+ AlternativeWithUtility(alt, math.exp(thisUtility))
```

Comment:
append to List is linear, so for the big input it can hurt performance. If order does not matter, use prepend:
https://docs.scala-lang.org/overviews/collections/performance-characteristics.html

Reply:
right. i need a way to differentiate between infinitely and non-infinitely-valued alternatives. i decided to prepend non-infinitely-valued alternatives, since that is more common (and more performant, as you mentioned), and to append the infinite ones. that way, i know, if there is at least one infinitely-valued alternative, i know where i can find it.